### PR TITLE
RFC: Sample architecture and onboarding documentation

### PR DESCRIPTION
**Pull request by ariard**
Ready for at least a high-level conceptual review.

For now, the sample is mostly a code skeleton with some commands functional (`connect`/`open`/...). Don't intend to use it beyond your local regtest.

### CODE REVIEW

**Comment by jkczyz** [comment]
Could you update the description to summarize the architecture?

File: src/init.rs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+use std::thread::sleep;
+use std::time::{Duration, SystemTime};
+
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::{Receiver, Sender};
+
+const FEE_PROPORTIONAL_MILLIONTHS: u32 = 10;
+const ANNOUNCE_CHANNELS: bool = false;
+
+#[tokio::main]
```
Comment (TheBlueMatt) [inline src/init.rs:44]: You may want to specify a threaded executor :)

File: src/init.rs [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+	// This sample node implementation communicate with the LDK command-line binary (`ldk-cli`)
+	// through a custom RPC protocol.
+
+	log_sample!(logger, "Starting rpc server thread...");
+
+	let outbound_rpc_server_connector = OutboundRPCServerConnector::new(outbound_peers_rpccmd, outbound_chanman_rpccmd, outbound_router_rpccmd);
+	let handles = setup_rpc_server(outbound_rpc_server_connector, ldk_port).await;
+	join_handles.push(handles);
+
+	loop {
```
Comment (TheBlueMatt) [inline src/init.rs:318]: Can't you just wait on all the join handles to finish?

File: src/sampled.rs [hunk 3]
```diff
@@ -0,0 +1,814 @@
...
+						let route = Route {
+							paths: vec![vec![hop]],	
+						};
+						channel_manager_rpc.send_payment(&route, PaymentHash(hash.into_inner()), &None).unwrap();
+						flush_msg_events!(outbound, channel_manager_rpc);
+					},
+					"invoice" => {
+						let amount = u64::from_str_radix(v_obj.get("amt").unwrap().as_str().unwrap(), 10).unwrap();
+						let payment_preimage = [1; 32];
+						//thread_rng().fill_bytes(&mut payment_preimage);
```
Comment (TheBlueMatt) [inline src/sampled.rs:212]: May want a TODO :)

**Review by TheBlueMatt — COMMENTED** [review]
Some comments from a quick once-over, I'll take a real look later.

File: ARCH.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+* `setup_rpc_server`, a simple RPC server handler to listen to `ldk-cli`
+
+All those components are running in their own threads and communicate between them using inter-threading
+handlers powered by `tokio::sync::mpsc`. As the communication channels provided by tokio are unidirectional,
+if the aforementioned components require bidirectional flows (e.g `ChannelManager` asking a route to
+`NetGraphMsgHandler`) a pair of inter-threading handlers are provided.
+
+# Why the sharing messages approach over the sharing memory one ?
+
+Long-term, it might be interesting to provide a multi-process LDK sample, better fitted for
```
Comment (TheBlueMatt) [inline ARCH.md:23]: Note that a big part of the motivation for this in Core is that its written in a memory-unsafe language. Without unsafe code, there's little reason to do this (absent compiler and other similar bugs). Cross-host, not just cross-process, stuff is cooler, though.

Comment (ariard) [reply ARCH.md:23]: Come on, playing with the security capabilities or privileges framework of your system is cool too. Or even control groups for intensive process like routing or throwing a load balancer between p2p stack and channel processing. For high-availability, high-security node, multi-process is way better :p

Comment (TheBlueMatt) [reply ARCH.md]: Lol, I mean I guess if you want capabilities enforcement that's cool. I still prefer multi-host, but whatever :)

**Pull request closed without merging**